plonky3/src/circuit_builder.rs

```diff
@@ -154,18 +198,33 @@ impl<'a, T: FieldElement> PowdrCircuit<'a, T> {
 impl<'a, T: FieldElement> BaseAir<Val> for PowdrCircuit<'a, T> {
     fn width(&self) -> usize {
         assert_eq!(self.analyzed.constant_count(), 0);
-        self.analyzed.commitment_count()
+        self.analyzed.commitment_count() + self.analyzed.publics_count()
     }
 
     fn preprocessed_trace(&self) -> Option<RowMajorMatrix<Val>> {
         panic!()
     }
 }
 
-impl<'a, T: FieldElement, AB: AirBuilder<F = Val>> Air<AB> for PowdrCircuit<'a, T> {
+impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val>> Air<AB> for PowdrCircuit<'a, T> {
     fn eval(&self, builder: &mut AB) {
         let matrix = builder.main();
+        let pi = builder.public_values();
+
+        // public constraints
+        let pi_moved = pi.into_iter()
+            .map(|itm| *itm)
+            .collect::<Vec<<AB as AirBuilderWithPublicValues>::PublicVar>>();
+        let local = matrix.row_slice(0);
+
+        // constraining Pi * (Ci - pub[i]) = 0
+        let mut pub_idx = 0;
+        for witness_col_idx in self.publics_idxs() {
```

Comment:
I think this is equivalent:
```suggestion
        for (pub_idx, witness_col_idx) in self.publics_idxs().iter().enumerate() {
```
Then we can remove the imperative `pub_idx` logic